### PR TITLE
Add support for 1d convolution in ttir and ttnn mlir dialects

### PR DESCRIPTION
**Pull request by ajakovljevicTT**
As described in issue https://github.com/tenstorrent/tt-mlir/issues/1342, ttnn does not currently have support for 1D convolution (bug opened https://github.com/tenstorrent/tt-metal/issues/15452). Hence, in talks with @LPanosTT, we agreed that it is ok to add reshapes before and after convolutions, so that we transform 1D convolution into shapes suitable for conv2d op in ttnn. 

This change is added in the `ConvolutionToConv2dPattern` in the 'TTIRToTTIRDecomposition', with different paths based on wheather the convolution is 2D or 1D.

The change was tested on convolution given in the issue above, which is:
```
module {
  func.func @main(%arg0: tensor<1x256x512xf32>, %arg1: tensor<1024x256x1xf32>, %arg2: tensor<1024xf32>) -> tensor<1x1024x512xf32> {
    %0 = stablehlo.convolution(%arg0, %arg1) dim_numbers = [b, f, 0]x[o, i, 0]->[b, f, 0], window = {stride = [1], pad = [[0, 0]], rhs_dilate = [1]} {batch_group_count = 1 : i64, feature_group_count = 1 : i64} : (tensor<1x256x512xf32>, tensor<1024x256x1xf32>) -> tensor<1x1024x512xf32>
    return %0 : tensor<1x1024x512xf32>
  }
}
```
And it works and runs on silicon.

### CODE REVIEW

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 1]
```diff
@@ -283,14 +287,149 @@ struct ConvolutionToConv2dPattern
...
+  ttir::ReshapeOp createReshapeOp(Location loc, Value tensor,
+                                  llvm::ArrayRef<int64_t> target_input_shape,
+                                  ::mlir::ArrayAttr constraints,
+                                  ConversionPatternRewriter &rewriter) const {
+    auto inputType = mlir::cast<RankedTensorType>(tensor.getType());
+
+    auto output = rewriter.create<tensor::EmptyOp>(
+        loc, llvm::ArrayRef<int64_t>(target_input_shape),
+        inputType.getElementType());
+    std::vector<int32_t> shapei32(target_input_shape.begin(),
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:299]: `llvm::SmallVector<int32_t, 2> shapei32(...)`

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 2]
```diff
@@ -283,14 +287,149 @@ struct ConvolutionToConv2dPattern
...
-  LogicalResult
-  matchAndRewrite(ttir::ConvolutionOp op, OpAdaptor adaptor,
-                  ConversionPatternRewriter &rewriter) const override {
+  ttir::ReshapeOp createReshapeOp(Location loc, Value tensor,
+                                  llvm::ArrayRef<int64_t> target_input_shape,
+                                  ::mlir::ArrayAttr constraints,
+                                  ConversionPatternRewriter &rewriter) const {
+    auto inputType = mlir::cast<RankedTensorType>(tensor.getType());
+
+    auto output = rewriter.create<tensor::EmptyOp>(
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:296]: Rename `output` to something else to signify that it's an op, since the type is `tensor::EmptyOp`, not an output `Value` of the op itself.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 3]
```diff
@@ -283,14 +287,149 @@ struct ConvolutionToConv2dPattern
...
+                              std::vector<bool> additionalIntegers) const {
+    llvm::SmallVector<bool, 4> newDenseArray(denseArrayAttr.asArrayRef());
+    for (bool boolean : additionalIntegers) {
+      newDenseArray.push_back(boolean);
+    }
+    return rewriter.getDenseBoolArrayAttr(newDenseArray);
+  }
 
-    if (failed(isConv2d(op))) {
+  LogicalResult matchAndRewrite1d(ttir::ConvolutionOp op, OpAdaptor adaptor,
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:332]: Can you add commentary outlining what's going on in this fn? It's hard to follow thru code only.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 4]
```diff
@@ -283,14 +287,149 @@ struct ConvolutionToConv2dPattern
...
+        rewriter, adaptor.getWeightDilationAttr(), std::vector<uint64_t>({1}));
+    mlir::DenseBoolArrayAttr newWindowReversalAttr =
+        addBooleansToDenseArrayAttr(rewriter, adaptor.getWindowReversalAttr(),
+                                    std::vector<bool>({false}));
+
+    auto convolutionLayout = adaptor.getConvolutionLayoutAttr();
+
+    // The additional spatial dimension is added at the and (3rd in 0 indexed
+    // array).
+    llvm::SmallVector<int64_t, 8> newInputSpatialDimensions(
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:385]: Why 8 values?

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 5]
```diff
@@ -283,14 +287,149 @@ struct ConvolutionToConv2dPattern
...
+    if (op.getConvolutionLayout().getInputSpatialDimensions()[0] != 2) {
       return failure();
     }
 
+    auto outputType =
+        mlir::cast<RankedTensorType>(adaptor.getOutput().getType());
+    llvm::ArrayRef<int64_t> outputShape = outputType.getShape();
+    llvm::SmallVector<int64_t, 8> newOutputShape(outputShape.begin(),
+                                                 outputShape.end());
+    newOutputShape.push_back(1);
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:344]: There are a lot of these pushes without any comments, I don't understand what they're for.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 6]
```diff
@@ -248,9 +253,8 @@ struct ConvolutionToConv2dPattern
       SPATIAL_DIM_WIDTH,
   };
 
-  LogicalResult isConv2d(ttir::ConvolutionOp op) const {
-
-    // Conv2d will have 2 spatial dimensions
+  LogicalResult isValidConv(ttir::ConvolutionOp op,
+                            uint32_t numSpatialDims) const {
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:257]: This method is kind of broken now. The intent is to check whether it's a valid conv, but it only compares spatial dims cnt vs the provided `numSpatialDims`, so if someone pushes a 3d conv thru, it will deem it valid. It's a strange pattern to have to provide the intended number of spatial dims to the fn. In my eyes, it makes more sense for the fn to check if the conv is 1d/2d vs something else.

Btw with this change you've overshadowed the more global `constexpr static uint32_t numSpatialDims = 2;` which was intended to signal that only 2d convs were valid.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 7]
```diff
@@ -283,14 +287,149 @@ struct ConvolutionToConv2dPattern
...
+        rewriter, adaptor.getWindowStridesAttr(), std::vector<uint64_t>({1}));
+    mlir::DenseI64ArrayAttr newPaddingAttr = addIntegersToDenseArrayAttr(
+        rewriter, adaptor.getPaddingAttr(), std::vector<uint64_t>({0, 0}));
+    mlir::DenseI64ArrayAttr newInputDilationAttr = addIntegersToDenseArrayAttr(
+        rewriter, adaptor.getInputDilationAttr(), std::vector<uint64_t>({1}));
+    mlir::DenseI64ArrayAttr newWeightDilationAttr = addIntegersToDenseArrayAttr(
+        rewriter, adaptor.getWeightDilationAttr(), std::vector<uint64_t>({1}));
+    mlir::DenseBoolArrayAttr newWindowReversalAttr =
+        addBooleansToDenseArrayAttr(rewriter, adaptor.getWindowReversalAttr(),
+                                    std::vector<bool>({false}));
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:379]: It seems like an overkill to have several methods that only add to a dense array attr. Also, you're creating whole vectors as well for single values, which is another overkill.

Maybe create `SmallVector`s from existing attr arrays, add needed values to it, and create `DenseI64ArrayAttr` based on that?

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 8]
```diff
@@ -283,14 +287,149 @@ struct ConvolutionToConv2dPattern
...
+    newWeightShape.push_back(1);
+
+    ttir::ReshapeOp reshapeInput =
+        createReshapeOp(op.getLoc(), adaptor.getInput(), newInputShape,
+                        op.getOperandConstraints(), rewriter);
+    ttir::ReshapeOp reshapeWeight =
+        createReshapeOp(op.getLoc(), adaptor.getWeight(), newWeightShape,
+                        op.getOperandConstraints(), rewriter);
+
+    mlir::DenseI64ArrayAttr newWindowsStridesAttr = addIntegersToDenseArrayAttr(
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:369]: `new` is not a great prefix, you use it for multiple ops here (e.g. reshape above also has `newOutputShape`), consider being more specific in naming.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 9]
```diff
@@ -229,6 +229,11 @@ struct ConvolutionToConv2dPattern
 public:
   using OpConversionPattern<ttir::ConvolutionOp>::OpConversionPattern;
 
+  ConvolutionToConv2dPattern(TypeConverter &typeConverter, MLIRContext *ctx)
+      : OpConversionPattern<ttir::ConvolutionOp>(typeConverter, ctx) {
+    setHasBoundedRewriteRecursion();
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:234]: Please add commentary explaining why this is necessary.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 6 above)
Comment (svuckovicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:257]: I guess we should think in terms of support, not validity. `isSupportedConv` would better convey the intent of the function after above-mentioned changes.

**Comment by ajakovljevicTT** [comment]
> Sorry for being late to take a look.
> 
> I think it might be better to make a “ttir.conv1d” operation, and a separate ConvolutionToConv1D pattern rewriter. This way the op is clearly differentiated in the IR. We can add the reshapes that insert the second spatial dim during lowering to TTNN. And if they ever do expose conv1d as a c++ api (which shouldn’t be hard) we can just bring that up between TTNN dialect and runtime.
> 
> I’m saying this because my original intent with ConvolutionToConv2D was to convert any Convolution with two spatial dimensions to a Conv2D op by inserting whatever transposes necessary on the input/output. This is semantically not a conv2D although it can be mimicked by one.
> 
> @svuckovicTT what do you think?

Thanks for looking into this @LPanosTT ! I have thought about the thing that you proposed when I first started implementing this, and had discussions with the TTIR dialect team (@sdjordjevicTT can chime in if has the time) about the best way to do this. We have settled on this solution, as, at least to me, there is no considerable benefit to making a `ttir.conv1d` currently, without the adequate ttnn/metal support. Furthermore, the conversion from ttir.convolution to ttnn.conv2d is non trivial (as can be seen from the `ConvolutionToConv2dPattern` in `lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp`), and most of that logic would need to be replicated to in the conv1d decomposition pass, only to be later deleted if ttnn folks expose the con1d C++ ttnn op. If you feel strongly about moving this conversion to TTIR->TTNN pass, we can discuss this further.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 6 above)
Comment (ajakovljevicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp]: With regards to change change of structure outlined in the comment written below, I have changed both of the validity/support checks (for `numSpatialDims` and otherwise) to live in the base `ConvolutionDecompositionPattern` class, which both `ConvolutionToConv2dPattern` and `Legalize1DConvolutionPattern` inherit from.
Please go through the changes regarding comments one more time and let me know if you think I missed something. Thanks!

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 3 above)
Comment (ajakovljevicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp]: I have added the comments at the header of a newly added pattern for 1d convolution. Please let me know if you think I should go into more details in the code itself.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 10]
```diff
@@ -224,34 +224,27 @@ static std::vector<TransposeDims> generateConvKernelTransposeIndices(
...
-      SPATIAL_DIM_HEIGHT,
-      SPATIAL_DIM_WIDTH,
-  };
+  LogicalResult
+  matchAndRewrite(ttir::ConvolutionOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override = 0;
 
-  LogicalResult isConv2d(ttir::ConvolutionOp op) const {
+protected:
+  LogicalResult isNDimensional(ttir::ConvolutionOp op,
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:237]: What's the value of using `LogicalResult` instead of `bool` for these utility functions?

File: lib/Dialect/TTNN/Transforms/TTNNLayout.cpp [hunk 11]
```diff
@@ -304,6 +318,11 @@ class TTNNLayoutDPSOperandsRewriter
       // TTNN Conv2d moves input, weight, and bias from host to device
       // itself. Inserting the ToLayoutOp on these operands is thus problematic.
       if (mlir::isa<ttir::Conv2dOp>(op.getOperation()) && !isResult) {
+        // For the weight input of the conv2d op, it specifically needs to be on
+        // host, so we create a host to_layout op.
+        if (operand.getOperandNumber() == 1) {
+          modified = changeLayoutToHost(op, operand, rewriter);
+        }
```
Comment (svuckovicTT) [inline lib/Dialect/TTNN/Transforms/TTNNLayout.cpp:325]: Please open an issue to track this and tag @sdjordjevicTT - this should be moved to workarounds pass eventually.

File: test/ttmlir/Conversion/StableHLOToTTIR/conv2d_op.mlir [hunk 12]
```diff
@@ -16,4 +16,24 @@ module @jit_convolution attributes {} {
...
+        pad = [[0, 0]],
+        rhs_dilate = [1]
+      } {
+        batch_group_count = 1 : i64,
+        feature_group_count = 1 : i64
+      } : (tensor<1x256x512xf32>, tensor<1024x256x1xf32>) -> tensor<1x1024x512xf32>
+    // CHECK: %1 = "ttir.convolution"(%arg0, %arg1, [[VAL0]])
+    // CHECK: batch_group_count = 1 : i64, convolution_layout = #ttir<convolution_layout input_batch = 0, input_feature = 1, input_spatial_dimensions = 2, kernel_output_feature = 0, kernel_input_feature = 1, kernel_spatial_dimensions = 2, output_batch = 0, output_feature = 1, output_spatial_dimensions = 2
+    // CHECK: padding = array<i64: 0, 0>, weight_dilation = array<i64: 1>, window_reversal = array<i1: false>, window_strides = array<i64: 1>
+    // CHECK: : (tensor<1x256x512xf32>, tensor<1024x256x1xf32>, [[TENSOR_SIZE]]) -> [[TENSOR_SIZE]]
```
Comment (svuckovicTT) [inline test/ttmlir/Conversion/StableHLOToTTIR/conv2d_op.mlir:36]: Can you move these above the `%0 = stablehlo.convolution...` line?

We usually write checks for an op _above_ the op itself.

Comment (svuckovicTT) [reply test/ttmlir/Conversion/StableHLOToTTIR/conv2d_op.mlir:36]: Same for the other test.

**Review by svuckovicTT — APPROVED** [review]

File: lib/Dialect/TTNN/Transforms/TTNNLayout.cpp (hunk 11 above)
Comment (ajakovljevicTT) [reply lib/Dialect/TTNN/Transforms/TTNNLayout.cpp]: Added the issue here: https://github.com/tenstorrent/tt-mlir/issues/1528

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 10 above)
Comment (ajakovljevicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp]: You are correct, there is no added value. Changed to `bool`.

File: test/ttmlir/Conversion/StableHLOToTTIR/conv2d_op.mlir (hunk 12 above)
Comment (ajakovljevicTT) [reply test/ttmlir/Conversion/StableHLOToTTIR/conv2d_op.mlir]: Fixed

**Review by LPanosTT — APPROVED** [review]
